### ISSUE TITLE
[Question] Issue with SignInManager 

### ISSUE THREAD

**Issue by dazinator** [aspnet/Home#1727]
TLDR: Why would SignInManager.IsSignedIn(User) be returning false for an authenticated User?

My set up is a little complex, but I don't want to skip any any detail in case it's important!

I have two websites, one is a legacy asp.net 3.5 application, and the other is an asp.net core app.

```
legacy.mydomain.com
aspcoreapp.mydomain.com
```

The user logs in at the legacy site, which produces an old style (.net 3.5) auth cookie, containing the old style forms authentication ticket (encrypted using AES with SHA1 hash validation)

In order to give single sign on accross the sites, I managed to implement a custom `ISecureDataFormat<AuthenticationTicket>` that I can use with the Cookie middleware in my asp.net core application. This custom format decrypts the legacy cookie data into a `FormsAuthenticationTicket` and then converts that into an `AuthenticationTicket` which it returns to asp.net core.

This works swimmingly.

```

  var legacyCookieFormat = new LegacyCookieTicketFormat(_DecryptionKeyText, _ValidationKeyText, principalProvider);

            var cookieOptions = new CookieAuthenticationOptions()
            {
                LoginPath = loginPath,
                AutomaticChallenge = true,
                AutomaticAuthenticate = true,
                AuthenticationScheme = "Cookies",
                TicketDataFormat = legacyCookieFormat,
                CookieName = "CookieA",
                ExpireTimeSpan = TimeSpan.FromMinutes(120)
            };
            
            // This blank domain is for running locally (localhost)
            cookieOptions.CookieDomain = "";
            app.UseCookieAuthentication(cookieOptions);

          
```

Next.. there are multiple tenants each using the sites with their own domain name.  I am registering a seperate authentication cookie per tenant domain. I implemented this like so:

```

             // Need to configure cookie authentication for each domain (per organisation)
            // This blank domain is for running locally
            cookieOptions.CookieDomain = "";
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = Domains.TenantADomain; // e.g *.tenantA.com
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = Domains.TenantBDomain; // e.g *.tenantB.com
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = Domains.TenantCDomain; // e.g *.tenantC.com
            app.UseCookieAuthentication(cookieOptions);        

```

This all works fantastic thus far.

But now I am trying to add Identity into the mix, with a custom store.
My custom store looks at existing database tables (in the legacy app).
I have implemented the necessary store classes and they all work swimmingly.

So here is what I do in `startup.cs` to register identity with the custom store:

```

 services.AddIdentity<MyUser, MySecurityRole>()
                .AddUserStore<MyUserStore>()
                .AddRoleStore<MyRoleStore>()
                .AddErrorDescriber<IdentityErrorDescriber>()
                .AddDefaultTokenProviders();

  services.AddScoped<IPasswordHasher<MyUser>, MyLegacyPasswordHasher>();

```

Note: I also register a custom password hasher, because I need to generate old style hashes that will validate against the existing legacy accounts.

Now with this in place:

When I log a user in using:

```

  var result = await _signInManager.PasswordSignInAsync(model.Email, model.Password, model.RememberMe, lockoutOnFailure: false);
                if (result.Succeeded)
                {
                    _logger.LogInformation(1, "User logged in.");
                    return RedirectToLocal(returnUrl);
                }
```

The log in code succeeds, and I can see my custom `UserClaimsPrincipalFactory` firing to construct a ClaimsPrincipal for identity.

```

  public class UserClaimsPrincipalFactory : IUserClaimsPrincipalFactory<MyUser>
    {
        public Task<ClaimsPrincipal> CreateAsync(MyUser user)
        {         

           // Do I have to use this name?
            ClaimsIdentity identity = new ClaimsIdentity("Microsoft.AspNet.Identity.Application");

           // adding some more claims here.     
           
            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, user.UserId.ToString()));     
            identity.AddClaim(new Claim(ClaimTypes.Email, user.Email));

            ClaimsPrincipal principal = new ClaimsPrincipal(identity);
            return Task.FromResult(principal);
        }
    }

```

Side Question: What is the significance of "Microsoft.AspNet.Identity.Application" for the claim name- could I use another name?

I am then successfully redirected back to the application and I am now logged in.. Well.. semi logged in!

I can see that User is being set to an authenticated claims principal, however `SignInManager.IsSignedIn(User)` always returns false.

Here is a screenshot of my debugger showing the current User, and the result from SignInManager

![image](https://cloud.githubusercontent.com/assets/3176632/17980175/f6dfb08a-6af5-11e6-89a1-bc12bb870f33.png)

I am confused as to what I need to change, or where I need to look to make SignInManager work correctly. I know this must boil down to some naming convention between authentication schemes or something but this stuff is really very confusing!

**Comment by dazinator** [comment]
Additional question: 
Registering Identity, and also registering cookie middleware directly. both give you cookie options.
Whats the correct way to handle having seperate cookies per domain? Without identity in the mix I did this:

```

            // This blank domain is for running locally
             cookieOptions.CookieDomain = "";
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = NavigateBetweenWebApps.ReachDomain;
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = NavigateBetweenWebApps.ShopSmartDomain;
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = NavigateBetweenWebApps.DisruptDomain;
            app.UseCookieAuthentication(cookieOptions);

            cookieOptions.CookieDomain = NavigateBetweenWebApps.FlexForceDomain;
            app.UseCookieAuthentication(cookieOptions);

```

No that I have added Identity, how do I correctly set it up?

```
// do i call this multiple times?
 services.AddIdentity<MyUser, MySecurityRole>((a) =>
            {
                // what do I do here?
                // a.Cookies.ApplicationCookie.?
            })

```

**Comment by wanton7** [comment]
Hi had exactly same problem. Fixed it with this.

    public class ApplicationUserPrincipalFactory : IUserClaimsPrincipalFactory<ApplicationUser>
    {
        public IdentityOptions Options { get; }

        public ApplicationUserPrincipalFactory(IOptions<IdentityOptions> optionsAccessor)
        {
            Options = optionsAccessor.Value;
        }

        public Task<ClaimsPrincipal> CreateAsync(ApplicationUser user)
        {
            var id = new ClaimsIdentity(Options.Cookies.ApplicationCookieAuthenticationScheme,
                Options.ClaimsIdentity.UserNameClaimType,
                Options.ClaimsIdentity.RoleClaimType);
            id.AddClaim(new Claim(Options.ClaimsIdentity.UserIdClaimType, user.Id));
            id.AddClaim(new Claim(Options.ClaimsIdentity.UserNameClaimType, user.UserName));

            return Task.FromResult(new ClaimsPrincipal(id));
        }
    }

**Comment by dazinator** [comment]
@wanton7  - cheers for that.

For anyone else hitting this, if you happy with the default claims and just want to add more, yo ucan derive from the default:

```

    public class UserClaimsPrincipalFactory : UserClaimsPrincipalFactory<User, SecurityRole>
    {
        public override async Task<ClaimsPrincipal> CreateAsync(User user)
        {

            var principal = await base.CreateAsync(user);

            ((ClaimsIdentity)principal.Identity)
                .AddClaims(new[] {
                    new Claim("SomeAdditionalClaim", user.SomeField.ToString())                 
                });
          
            return principal;
        }

        public UserClaimsPrincipalFactory(UserManager<User> userManager, RoleManager<SecurityRole> roleManager, IOptions<IdentityOptions> optionsAccessor) : base(userManager, roleManager, optionsAccessor)
        {

        }
    }


```
